### PR TITLE
Use new Global Configs

### PR DESCRIPTION
**Pull request by Romazes**
#### Description

- Use Globals for credentials access for subscription validation.
This follows https://github.com/QuantConnect/Lean/pull/7796
- Remove validation of ApiKey from `GetHistory()`. It is checked in initialize method.
- Handle history provider returning null
  - This follows https://github.com/QuantConnect/Lean/pull/7804
- Handle data downloader provider returning null
  - This follows https://github.com/QuantConnect/Lean/pull/7812

#### Related Issue

N/A

#### Motivation and Context

Maintain a uniform code format like in others repos.

#### Requires Documentation Change

N/A

#### How Has This Been Tested?

N/A

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: QuantConnect.Polygon.Tests/PolygonDataDownloaderTests.cs [hunk 1]
```diff
@@ -60,17 +60,22 @@ public void DownloadsHistoricalData(Symbol symbol, Resolution resolution, TimeSp
...
             var request = PolygonHistoryTests.CreateHistoryRequest(symbol, resolution, tickType, period);
 
             var parameters = new DataDownloaderGetParameters(symbol, resolution, request.StartTimeUtc, request.EndTimeUtc, tickType);
-            var data = _downloader.Get(parameters).ToList();
+            var data = _downloader.Get(parameters)?.ToList();
+
+            if (data == null)
+            {
+                Assert.Pass("History returns null result");
+            }
```
Comment (jhonabreul) [inline QuantConnect.Polygon.Tests/PolygonDataDownloaderTests.cs:74]: I think you sould add test cases that return null instead of just letting the tests pass

File: QuantConnect.Polygon.Tests/PolygonHistoryTests.cs [hunk 2]
```diff
@@ -148,7 +148,12 @@ public void GetsIndexHistoricalData(Resolution resolution, TimeSpan period, Tick
         {
             var symbol = Symbol.Create("SPX", SecurityType.Index, Market.USA);
             var requests = new List<HistoryRequest> { CreateHistoryRequest(symbol, resolution, tickType, period) };
-            var history = _historyProvider.GetHistory(requests, TimeZones.Utc).ToList();
+            var history = _historyProvider.GetHistory(requests, TimeZones.Utc)?.ToList();
+
+            if (history == null)
+            {
+                Assert.Pass("History returns null result");
+            }
```
Comment (jhonabreul) [inline QuantConnect.Polygon.Tests/PolygonHistoryTests.cs:156]: Same comment as for the data downloader tests

File: QuantConnect.Polygon.Tests/PolygonHistoryTests.cs [hunk 3]
```diff
@@ -223,9 +228,9 @@ public void ReturnsEmptyForUnsupportedSecurityTypeResolutionOrTickType(Symbol sy
         {
             using var historyProvider = new TestPolygonHistoryProvider();
             var request = CreateHistoryRequest(symbol, resolution, tickType, TimeSpan.FromDays(100));
-            var history = historyProvider.GetHistory(request).ToList();
+            var history = historyProvider.GetHistory(request)?.ToList();
 
-            Assert.That(history, Is.Empty);
+            Assert.IsNull(history);
```
Comment (jhonabreul) [inline QuantConnect.Polygon.Tests/PolygonHistoryTests.cs:233]: Since returning null instead of empty now, the test name should be changed

File: QuantConnect.Polygon/PolygonDataDownloader.cs [hunk 4]
```diff
@@ -65,7 +66,7 @@ public IEnumerable<BaseData> Get(DataDownloaderGetParameters parameters)
 
             if (endUtc < startUtc)
             {
-                yield break;
+                return null;
```
Comment (jhonabreul) [inline QuantConnect.Polygon/PolygonDataDownloader.cs:69]: Could return empty instead of null for this specific case

File: QuantConnect.Polygon/PolygonDataDownloader.cs [hunk 5]
```diff
@@ -74,35 +75,64 @@ public IEnumerable<BaseData> Get(DataDownloaderGetParameters parameters)
...
-                    yield return data;
+                    return null;
                 }
+
+                return GetBaseData(historyData);
             }
-            else
+        }
+
+        private IEnumerable<BaseData>? GetBaseData(Symbol symbol, DateTime startUtc, DateTime endUtc, Type dataType, Resolution resolution, SecurityExchangeHours exchangeHours, DateTimeZone dataTimeZone, TickType tickType)
```
Comment (jhonabreul) [inline QuantConnect.Polygon/PolygonDataDownloader.cs:96]: I think it is a bit unclear that this method does by just looking at its name an signature. Perhaps `GetCanonicalOptionHistory` would be clearer.

File: QuantConnect.Polygon/PolygonDataDownloader.cs [hunk 6]
```diff
@@ -74,35 +75,64 @@ public IEnumerable<BaseData> Get(DataDownloaderGetParameters parameters)
...
+                {
+                    dataQueue.Add(data);
+                }
+            })).ContinueWith(_ =>
+            {
+                dataQueue.CompleteAdding();
+            });
+
+            // Validate: data is not null, we have gotten anything at least
+            if (!dataQueue.TryTake(out _, TimeSpan.FromSeconds(10)))
```
Comment (jhonabreul) [inline QuantConnect.Polygon/PolygonDataDownloader.cs:123]: Careful here, if history was not null, you would be consuming one data point/bar that would be lost. You would need some other thing for this check, like having two events, `historyNullEvent` and `historyNotNull` that are set by the task. Then here you wait for both of them and depending on the event that is actually triggered you either return null or `dataQueue.GetConsumingEnumerable()`.

File: QuantConnect.Polygon/PolygonDataDownloader.cs [hunk 7]
```diff
@@ -74,35 +75,64 @@ public IEnumerable<BaseData> Get(DataDownloaderGetParameters parameters)
...
+            {
+                return null;
+            }
+
+            return GetBaseData(dataQueue.GetConsumingEnumerable());
+        }
+
+        private IEnumerable<BaseData> GetBaseData(IEnumerable<BaseData> baseData)
```
Comment (jhonabreul) [inline QuantConnect.Polygon/PolygonDataDownloader.cs:131]: You can remove this method, just return  `dataQueue.GetConsumingEnumerable()` and `historyData` in the methods above.

File: QuantConnect.Polygon/PolygonDataDownloader.cs [hunk 8]
```diff
@@ -13,10 +13,11 @@
  * limitations under the License.
  */
 
-using QuantConnect.Configuration;
+using NodaTime;
```
Comment (jhonabreul) [inline QuantConnect.Polygon/PolygonDataDownloader.cs:16]: Is this used?

File: QuantConnect.Polygon/PolygonDataProvider.cs [hunk 9]
```diff
@@ -175,16 +179,13 @@ public void SetJob(LiveNodePacket job)
...
 
-            _apiKey = apiKey;
-
-            var maxSubscriptionsPerWebSocket = 0;
             if (!job.BrokerageData.TryGetValue("polygon-max-subscriptions-per-websocket", out var maxSubscriptionsPerWebSocketStr) ||
-                !int.TryParse(maxSubscriptionsPerWebSocketStr, out maxSubscriptionsPerWebSocket))
+                !int.TryParse(maxSubscriptionsPerWebSocketStr, out var maxSubscriptionsPerWebSocket))
             {
                 maxSubscriptionsPerWebSocket = -1;
             }
```
Comment (jhonabreul) [inline QuantConnect.Polygon/PolygonDataProvider.cs:186]: This I think is a small bug. The default value for `maxSubscriptionsPerWebSocket` is -1. Could you please adapt it so that is `job.BrokerageData` doesn't have it or it can't be parsed as an integer, the default value is -1 always and not 0?

**Review by jhonabreul — CHANGES_REQUESTED** [review]
Thanks! Leaving some comments

File: QuantConnect.Polygon.Tests/PolygonDataDownloaderTests.cs (hunk 1 above)
Comment (Romazes) [reply QuantConnect.Polygon.Tests/PolygonDataDownloaderTests.cs:74]: caught red-handed 🙄

Comment (Romazes) [reply QuantConnect.Polygon.Tests/PolygonDataDownloaderTests.cs]: Seperate in two methods. 
use DRY

File: QuantConnect.Polygon/PolygonDataDownloader.cs (hunk 4 above)
Comment (Romazes) [reply QuantConnect.Polygon/PolygonDataDownloader.cs]: return `return Enumerable.Empty<BaseData>();` instead of `null`.

File: QuantConnect.Polygon/PolygonDataDownloader.cs (hunk 5 above)
Comment (Romazes) [reply QuantConnect.Polygon/PolygonDataDownloader.cs]: Thanks for tip.

File: QuantConnect.Polygon/PolygonDataDownloader.cs (hunk 7 above)
Comment (Romazes) [reply QuantConnect.Polygon/PolygonDataDownloader.cs]: Great. It is working 😋

File: QuantConnect.Polygon/PolygonDataDownloader.cs (hunk 8 above)
Comment (Romazes) [reply QuantConnect.Polygon/PolygonDataDownloader.cs:16]: image

File: QuantConnect.Polygon/PolygonDataDownloader.cs (hunk 6 above)
Comment (Romazes) [reply QuantConnect.Polygon/PolygonDataDownloader.cs]: I have refactored in different way.

**Review by jhonabreul — APPROVED** [review]

**Review by Martin-Molinero — APPROVED** [review]
Thank you 👍
